### ISSUE TITLE
`indexing_slicing` should not trigger if the `get` method expects a different type

### ISSUE THREAD

**Issue by ronnodas** [rust-lang/rust-clippy#16384]
### Summary

As implemented in #12488, the `indexing_slicing` lint checks if there's a `get` method before suggesting using it. However, it doesn't check if that method can be used with the type being used as the index.

### Lint Name

indexing_slicing

### Reproducer

I tried this code (simplified from the [`linearize`](https://crates.io/crates/linearize) docs):

```rust
use linearize::{Linearize, static_map};

#[derive(Linearize)]
enum Keys {
    A,
    B,
}

fn main() {
    let map = static_map! {
        Keys::A => "a",
        Keys::B => "b",
    };
    assert_eq!(map[Keys::A], "a");
}
```

I saw this happen:

```
    = help: consider using `.get(n)` or `.get_mut(n)` instead
    = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.92.0/index.html#indexing_slicing
```
This is because `StaticMap<Keys, &str>` derefs to `[&str]`, which has a `get(usize)` method. However, this cannot be used to index by something of type `Keys`, so I expected the lint not to trigger.

### Version

```text

```
